src/browser/web_request_handler.ts

```diff
@@ -0,0 +1,145 @@
+/*
+Copyright 2017 OpenFin Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+    Intercept and modify the contents of a request at various stages of its lifetime, based on
+    https://github.com/openfin/runtime/blob/develop/docs/api/web-request.md
+
+    v1: handler for onBeforeSendHeaders
+ */
+
+const coreState = require('./core_state');
+const electronApp = require('electron').app;
+const {session, webContents} = require('electron');
+
+import * as Shapes from '../shapes';
+
+const moduleName: string = 'WebRequestHandlers';  // for logging
+
+// passed to webRequest.onBeforeSendHeaders
+interface RequestDetails {
+    id: number;
+    url: string;
+    method: string;
+    resourceType: string;
+    requestHeaders: any;
+    renderProcessId?: number; // not set if the request is not associated with a window
+    renderFrameId?: number;
+}
+
+// passed to callback of webRequest.onBeforeSendHeaders
+interface HeadersResponse {
+    cancel: boolean;
+    requestHeaders?: any;
+}
+
+interface AppUuidFilterUrlMap {
+    [uuid: string]: Shapes.WebRequestHeaderConfig[];  // app UUID => custom header settings
+}
+const filterMap: AppUuidFilterUrlMap = {};
+
+interface UrlHeaderMap {
+    [urlPattern: string]: Shapes.WebRequestHeader[];  // URL pattern => list of headers
+}
+let headerMap: UrlHeaderMap = {};  // reset for every app started and closed
+
+function matchUrlPatterns(url: string, config: Shapes.WebRequestHeaderConfig): boolean {
+    let match: boolean = false;
+    if (config.urlPatterns && config.urlPatterns.length > 0) {
+            match = electronApp.matchesURL(url, config.urlPatterns);
+    }
```

Comment:
Suggestion: Allow `urlPatterns` to be optional and return `true` here if omitted, meaning headers are applied to all windows app-wide (essentially defaulting to `['<all_urls>']`):
```js
    let urlPatternsSpecified = config.urlPatterns && config.urlPatterns.length > 0;
    let match = !urlPatternsSpecified || electronApp.matchesURL(url, config.urlPatterns);
```